shadowsocks-csharp/Controller/Service/GeositeUpdater.cs

```diff
@@ -73,31 +93,83 @@ public static void UpdatePACFromGeosite(Configuration config)
                 geositeUrl = config.geositeUrl;
             }
             logger.Info($"Checking Geosite from {geositeUrl}");
-            WebClient http = new WebClient();
+
+            // use System.Net.Http.HttpClient to download GeoSite db.
+            // NASTY workaround: new HttpClient every update
+            // because we can't change proxy on existing socketsHttpHandler instance
+            socketsHttpHandler = new SocketsHttpHandler();
+            httpClient = new HttpClient(socketsHttpHandler);
             if (config.enabled)
             {
-                http.Proxy = new WebProxy(
+                socketsHttpHandler.UseProxy = true;
+                socketsHttpHandler.Proxy = new WebProxy(
                     config.isIPv6Enabled
                     ? $"[{IPAddress.IPv6Loopback}]"
                     : IPAddress.Loopback.ToString(),
                     config.localPort);
             }
-            http.DownloadDataCompleted += (o, e) =>
+
+            try
             {
-                try
+                // download checksum first
+                var geositeSha256sum = await httpClient.GetStringAsync(geositeSha256sumUrl);
+                geositeSha256sum = geositeSha256sum.Substring(0, 64).ToUpper();
+                logger.Info($"Got Sha256sum: {geositeSha256sum}");
+                // compare downloaded checksum with local geositeDB
+                byte[] localDBHashBytes = mySHA256.ComputeHash(geositeDB);
+                string localDBHash = BitConverter.ToString(localDBHashBytes).Replace("-", String.Empty);
+                logger.Info($"Local Sha256sum: {localDBHash}");
+                // if already latest
+                if (geositeSha256sum == localDBHash)
+                {
+                    logger.Info("Local GeoSite DB is already the latest.");
+                    return;
+                }
+
+                // not latest. download new DB
+                var downloadedBytes = await httpClient.GetByteArrayAsync(geositeUrl);
+
+                // verify sha256sum
+                byte[] downloadedDBHashBytes = mySHA256.ComputeHash(downloadedBytes);
+                string downloadedDBHash = BitConverter.ToString(downloadedDBHashBytes).Replace("-", String.Empty);
+                logger.Info($"Actual Sha256sum: {downloadedDBHash}");
+                if (geositeSha256sum != downloadedDBHash)
                 {
-                    File.WriteAllBytes(DATABASE_PATH, e.Result);
-                    LoadGeositeList();
+                    logger.Info("Sha256sum mismatch. Updating aborted.");
```

Comment:
"Sha256sum verification: FAIL. File is corrupted. Abort the update."

Reply:
👌 Refined the logging messages in b3e15fc.